allennlp/training/util.py

```diff
@@ -122,18 +123,85 @@ def str_to_time(time_str: str) -> datetime.datetime:
+                         cache_prefix: str = None) -> Dict[str, Iterable[Instance]]:
     """
     Load all the datasets specified by the config.
+
+    Parameters
+    ----------
+    params : ``Params``
+    cache_directory : ``str``, optional
+        If given, we will instruct the ``DatasetReaders`` that we construct to cache their
+        instances in this location (or read their instances from caches in this location, if a
+        suitable cache already exists).  This is essentially a `base` directory for the cache, as
+        we will additionally add the ``cache_prefix`` to this directory, giving an actual cache
+        location of ``cache_directory + cache_prefix``.
+    cache_prefix : ``str``, optional
+        This works in conjunction with the ``cache_directory``.  The idea is that the
+        ``cache_directory`` contains caches for all different parameter settings, while the
+        ``cache_prefix`` captures a specific set of parameters that led to a particular cache file.
+        That is, if you change the tokenization settings inside your ``DatasetReader``, you don't
+        want to read cached data that used the old settings.  In order to avoid this, we compute a
+        hash of the parameters used to construct each ``DatasetReader`` and use that as a "prefix"
+        to the cache files inside the base ``cache_directory``.  So, a given ``input_file`` would
+        be cached essentially as ``cache_directory + cache_prefix + input_file``, where you specify
+        a ``cache_directory``, the ``cache_prefix`` is based on the dataset reader parameters, and
+        the ``input_file`` is whatever path you provided to ``DatasetReader.read()``.  In order to
+        allow you to give recognizable names to these prefixes if you want them, you can manually
+        specify the ``cache_prefix``.  Note that in some rare cases this can be dangerous, as we'll
+        use the `same` prefix for both train and validation dataset readers.
     """
-    dataset_reader = DatasetReader.from_params(params.pop('dataset_reader'))
-    validation_dataset_reader_params = params.pop("validation_dataset_reader", None)
+    dataset_reader_params = params.pop('dataset_reader')
+    validation_dataset_reader_params = params.pop('validation_dataset_reader', None)
+    if cache_directory:
+        if not cache_directory.endswith('/'):
+            cache_directory += '/'
+
+        # We need to compute the parameter hash before the parameters get destroyed when they're
+        # passed to `DatasetReader.from_params`.
+        if not cache_prefix:
+            cache_prefix = _dataset_reader_param_hash(dataset_reader_params)
+            if validation_dataset_reader_params:
+                validation_cache_prefix = _dataset_reader_param_hash(validation_dataset_reader_params)
+            else:
+                validation_cache_prefix = cache_prefix
+        else:
+            validation_cache_prefix = cache_prefix
```

Comment:
1. I feel like the details of how the cache gets setup are ancillary to this function, so probably most of this "set up the cache" logic should be pulled out into its own function?
2. all this manual adding of slashes feels really clunky, I wonder if you could avoid it using either `os.path.join` or `pathlib.Path`?
3. similarly, all of the path creation using `+` feels clunky. 

Couldn't you do something like

```
cache_path = pathlib.Path(cache_directory) / cache_prefix
validation_cache_path = ...
```

and then you don't have to do all the addition?

Reply:
I believe I fixed all of this.